[cmake] microhttpd: only link with libraries

Rechi opened the pull request:
## Description
pkg-config adds the static libraries to libraries if microhttpd is a static lib

## Motivation and Context
`pkg-config --libs --static libmicrohttpd` on Debian lists `-lunistring`, but `libunistring.so` isn't installed.
It is packaged in the `libunistring-dev` package, and is only a symlink to the real versioned libs. Those are installed correctly, so not a packaging bug from Debian.

## How Has This Been Tested?
compiled for linux and osx

## Types of change
- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [x] All new and existing tests passed

Rechi commented:
@wsnipex ping

wsnipex commented on cmake/modules/FindMicroHttpd.cmake:
```diff
@@ -35,17 +35,7 @@ if(MICROHTTPD_FOUND)
-  else()
-    list(APPEND MICROHTTPD_LIBRARIES ${PC_MICROHTTPD_STATIC_LIBRARIES})
-  endif()
+  list(APPEND MICROHTTPD_LIBRARIES ${PC_MICROHTTPD_LIBRARIES})
```
shouldn't this be PC_MICROHTTPD_STATIC_LIBRARIES?

Rechi replied:
Adding `PC_MICROHTTPD_STATIC_LIBRARIES` unconditionally doesn't work, see PR description.

I thought it is a special pkg-config feature, that for static libs the private libs are also included, but it is caused by the way pkg-config is build in depends (`
--enable-indirect-deps`).
https://github.com/xbmc/xbmc/blob/ce136c8bc4a3a0af3f24752f2a7630d359ae10d6/tools/depends/native/pkg-config-native/Makefile#L19

I'm not sure if we should rely on this. @wsnipex what do you think?

wsnipex replied:
probably not, as it will break with static libs outside of our depends.

Rechi replied:
If one uses a toolchain with static libs, shouldn't then pkg-config also be build with `
--enable-indirect-deps`? Otherwise building most of libraries will fail anyway.

wsnipex replied:
I don't see why this would be mandatory, so we can't rely on that in general. It's ok for me for depends builds.

We should just check if we found libmicrohttpd.a and only then use  PC_MICROHTTPD_STATIC_LIBRARIES.
Something like https://github.com/xbmc/xbmc/blob/master/cmake/modules/FindFFMPEG.cmake#L182-L190

Rechi replied:
@wsnipex done

wsnipex commented on cmake/modules/FindMicroHttpd.cmake:
```diff
@@ -35,16 +35,8 @@ if(MICROHTTPD_FOUND)
-  else()
-    list(APPEND MICROHTTPD_LIBRARIES ${PC_MICROHTTPD_STATIC_LIBRARIES})
+  if(${MICROHTTPD_LIBRARY} MATCHES ".+\.a$" AND PC_MICROHTTPD_STATIC_LDFLAGS)
+    list(APPEND MICROHTTPD_LIBRARIES ${PC_MICROHTTPD_STATIC_LDFLAGS})
```
This still lists unistring, so it results in essentially the same behavior as without this PR for depends builds.

Rechi replied:
This is not for fixing depends builds, it's for fixing builds with shared system libs.

wsnipex replied:
Now I'm confused, since I'm not aware that they are broken. PPA builds have been fine for a long time.
I also took another look and even though unistring is referenced in the .pc file, cmake seems to filter this out:
>PC_MICROHTTPD_STATIC_LDFLAGS:INTERNAL=-lmicrohttpd;-lgnutls;-lgmp;/usr/lib/x86_64-linux-gnu/libunistring.so;-lidn2;-lhogweed;-lgmp;-lnettle;-ltasn1;-lp11-kit;-lz
PC_MICROHTTPD_STATIC_LDFLAGS_OTHER:INTERNAL=/usr/lib/x86_64-linux-gnu/libunistring.so
**PC_MICROHTTPD_STATIC_LIBRARIES:INTERNAL=microhttpd;gnutls;gmp;idn2;hogweed;gmp;nettle;tasn1;p11-kit;z**

Rechi replied:
### system libs
```
+ pkg-config --libs libmicrohttpd
-lmicrohttpd
+ pkg-config --libs-only-l libmicrohttpd
-lmicrohttpd
+ pkg-config --libs-only-L libmicrohttpd

+ pkg-config --libs-only-other libmicrohttpd

+ pkg-config --libs --static libmicrohttpd
-lmicrohttpd -lgnutls -lgmp -lunistring -lidn2 -latomic -lhogweed -lgmp -lnettle -ltasn1 -lp11-kit
+ pkg-config --libs-only-l --static libmicrohttpd
-lmicrohttpd -lgnutls -lgmp -lunistring -lidn2 -latomic -lhogweed -lgmp -lnettle -ltasn1 -lp11-kit
+ pkg-config --libs-only-L --static libmicrohttpd

+ pkg-config --libs-only-other --static libmicrohttpd

```
```
PC_MICROHTTPD_STATIC_LDFLAGS:INTERNAL=-lmicrohttpd;-lgnutls;-lgmp;-lunistring;-lidn2;-latomic;-lhogweed;-lgmp;-lnettle;-ltasn1;-lp11-kit
PC_MICROHTTPD_STATIC_LIBRARIES:INTERNAL=microhttpd;gnutls;gmp;unistring;idn2;atomic;hogweed;gmp;nettle;tasn1;p11-kit
```

### depends
```
+ pkg-config --libs libmicrohttpd
-L<PREFIX>lib -lmicrohttpd
+ pkg-config --libs-only-l libmicrohttpd
-lmicrohttpd
+ pkg-config --libs-only-L libmicrohttpd
-L<PREFIX>lib
+ pkg-config --libs-only-other libmicrohttpd

+ pkg-config --libs --static libmicrohttpd
-L<PREFIX>lib -lmicrohttpd -lgcrypt -lgpg-error -lgnutls -lz -lgmp -lhogweed -lgmp -lnettle
+ pkg-config --libs-only-l --static libmicrohttpd
-lmicrohttpd -lgcrypt -lgpg-error -lgnutls -lz -lgmp -lhogweed -lgmp -lnettle
+ pkg-config --libs-only-L --static libmicrohttpd
-L<PREFIX>lib
+ pkg-config --libs-only-other --static libmicrohttpd

```

I don't know what's different on your system, but linking with `PC_MICROHTTPD_STATIC_LIBRARIES` if microhttpd is a shared library is always wrong.

wsnipex replied:
I agree it's unexpected, but the thing is, it's not only my system, it also works fine on launchpad. And since launchpad uses completely clean ubuntu, this must be something on your end.
I'd like to find out what.

But using PC_MICROHTTPD_STATIC_LDFLAGS breaks it for me, because now unistring is not filtered anymore

Rechi replied:
If you are using the *shared* system library `PC_MICROHTTPD_STATIC_LDFLAGS` won't be added to `MICROHTTPD_LIBRARIES`, because of the `.a` check you requested to add. If you want, I can replace `PC_MICROHTTPD_STATIC_LDFLAGS` with `PC_MICROHTTPD_STATIC_LIBRARIES`.

wsnipex replied:
yes, plz change to PC_MICROHTTPD_STATIC_LIBRARIES

Rechi replied:
done

Rechi committed:
[cmake] microhttpd: only link with libraries

pkg-config adds the static libraries to libraries if microhttpd is a static lib

Rechi committed:
[depends] microhttpd: remove build & install hacks
